File: hspasov-web-server-cpp/logger.cpp
#include <list>
#include <map>
#include <unistd.h>
#include <stdlib.h>
#include <execinfo.h>
#include "rapidjson/document.h"
#include "logger.hh"
#include "error_log_fields.hh"
#include "access_log_fields.hh"
#include "config.hh"
#include "web_server_utils.hh"
#include "err_log_lvl.hh"

int Logger::access_log_fd = -1;
std::map<std::string, bool> Logger::selected_error_log_fields;
std::map<std::string, bool> Logger::selected_access_log_fields;

void Logger::init_logger () {
  const std::list<std::string> allowed_error_log_fields = {
      "pid",
      "timestamp",
      "level",
      "context",
      "var_name",
      "var_value",
      "msg",
    };

  const std::list<std::string> allowed_access_log_fields = {
    "pid",
    "timestamp",
    "remote_addr",
    "req_line",
    "user_agent",
    "status_code",
    "content_length"
  };

  for (
    std::list<std::string>::const_iterator it = allowed_error_log_fields.begin();
    it != allowed_error_log_fields.end();
    ++it) {

    selected_error_log_fields[*it] = false;
  }

  for (
    std::list<std::string>::const_iterator it = allowed_access_log_fields.begin();
    it != allowed_access_log_fields.end();
    ++it) {

    selected_access_log_fields[*it] = false;
  }

  for (
    rapidjson::Value::ConstValueIterator it = Config::config["error_log_fields"].GetArray().Begin();
    it != Config::config["error_log_fields"].GetArray().End();
    ++it) {

    Logger::selected_error_log_fields[it->GetString()] = true;
  }

  for (
    rapidjson::Value::ConstValueIterator it = Config::config["access_log_fields"].GetArray().Begin();
    it != Config::config["access_log_fields"].GetArray().End();
    ++it) {

    Logger::selected_access_log_fields[it->GetString()] = true;
  }
}

void Logger::init_access_log () {

}

void Logger::close_access_log () {

}

void Logger::error (const error_log_fields& fields) {
  if (fields.level <= Config::config["error_log_level"].GetInt()) {
    std::list<std::string> fields_list;

    if (Logger::selected_error_log_fields.at("pid")) {
      fields_list.push_back(std::to_string(getpid()));
    }

    if (Logger::selected_error_log_fields.at("timestamp")) {
      fields_list.push_back(web_server_utils::get_current_time());
    }

    if (Logger::selected_error_log_fields.at("level")) {
      std::map<err_log_lvl, std::string> err_log_lvl_str = {
        { ERROR, "ERROR" },
        { INFO, "INFO" },
        { WARNING, "WARNING" },
        { DEBUG, "DEBUG" },
      };

      fields_list.push_back(err_log_lvl_str[fields.level]);
    }

    if (Logger::selected_error_log_fields.at("context")) {
      void* backtrace_points[2];
      int entries = backtrace(backtrace_points, 2);
      char** backtrace_readables = backtrace_symbols(backtrace_points, entries);

      // 0 is current function, 1 is caller function
      std::string result(backtrace_readables[1]);

      free(backtrace_readables);

      fields_list.push_back(result);
    }

    if (Logger::selected_error_log_fields.at("var_name")) {
      if (fields.var_name.length() > 0) {
        fields_list.push_back(fields.var_name);
      } else {
        fields_list.push_back(Config::config["error_log_empty_field"].GetString());
      }
    }

    if (Logger::selected_error_log_fields.at("var_value")) {
      if (fields.var_value.length() > 0) {
        fields_list.push_back(fields.var_value);
      } else {
        fields_list.push_back(Config::config["error_log_empty_field"].GetString());
      }
    }

    if (Logger::selected_error_log_fields.at("msg")) {
      if (fields.msg.length() > 0) {
        fields_list.push_back(fields.msg);
      } else {
        fields_list.push_back(Config::config["error_log_empty_field"].GetString());
      }
    }

    for (
      std::list<std::string>::iterator it = fields_list.begin();
      it != fields_list.end();
      ++it) {

      if (it != fields_list.begin()) {
        std::cerr << Config::config["error_log_field_sep"].GetString();
      }

      std::cerr << *it;
    }

    std::cerr << std::endl;
  }
}

void Logger::access (const access_log_fields& fields) {
  if (Config::config["access_log_enabled"].GetBool()) {
    if (!web_server_utils::is_fd_open(Logger::access_log_fd)) {
      struct error_log_fields f = { ERROR };
      f.msg = "Attempt to write in uninitialized access log file";
      Logger::error(f);
    } else {
      std::list<std::string> fields_list;

      if (Logger::selected_access_log_fields.at("pid")) {
        fields_list.push_back(std::to_string(getpid()));
      }

      if (Logger::selected_access_log_fields.at("timestamp")) {
        fields_list.push_back(web_server_utils::get_current_time());
      }

      if (Logger::selected_access_log_fields.at("remote_addr")) {
        if (fields.remote_addr.length() > 0) {
          fields_list.push_back(fields.remote_addr);
        } else {
          fields_list.push_back(Config::config["access_log_empty_field"].GetString());
        }
      }

      if (Logger::selected_access_log_fields.at("req_line")) {
        if (fields.req_line.length() > 0) {
          fields_list.push_back(fields.req_line);
        } else {
          fields_list.push_back(Config::config["access_log_empty_field"].GetString());
        }
      }

      if (Logger::selected_access_log_fields.at("user_agent")) {
        if (fields.user_agent.length() > 0) {
          fields_list.push_back(fields.user_agent);
        } else {
          fields_list.push_back(Config::config["access_log_empty_field"].GetString());
        }
      }

      if (Logger::selected_access_log_fields.at("status_code")) {
        if (fields.status_code.length() > 0) {
          fields_list.push_back(fields.status_code);
        } else {
          fields_list.push_back(Config::config["access_log_empty_field"].GetString());
        }
      }

      if (Logger::selected_access_log_fields.at("content_length")) {
        if (fields.content_length.length() > 0) {
          fields_list.push_back(fields.content_length);
        } else {
          fields_list.push_back(Config::config["access_log_empty_field"].GetString());
        }
      }
    }
  }
}

File: hspasov-web-server-node/logger.js
'use strict';

const { createWriteStream } = require('fs');
const { assert, isObject } = require('./web_server_utils.js');
const { CONFIG } = require('./config.js');

const errorLogLevels = {
  ERROR: 1,
  INFO: 2,
  WARNING: 3,
  DEBUG: 4,
};

const accessLog = createWriteStream(CONFIG.access_log, {
  flags: 'a', // open for appending, create if not exists
  encoding: 'utf-8',
  autoClose: true,
  mode: 0o666,
});

let accessLogAvailable = false;

accessLog.on('close', () => {
  error(errorLogLevels.DEBUG, { msg: 'closed access log' });
  accessLogAvailable = false;
});

accessLog.on('error', (e) => {
  error(errorLogLevels.DEBUG, { msg: 'access log error', var_name: 'error', var_value: e });
  accessLogAvailable = false;
});

accessLog.on('open', (fd) => {
  error(errorLogLevels.INFO, { msg: 'access log opened', var_name: 'fd', var_value: fd });
  accessLogAvailable = true;
});

const error = (level, fields) => {
  assert(Object.values(errorLogLevels).includes(level));
  assert(isObject(fields));

  if (level <= CONFIG.error_log_level) {
    const fieldsList = [];

    if (CONFIG.error_log_fields.includes('pid')) {
      fieldsList.push(process.pid);
    }

    if (CONFIG.error_log_fields.includes('timestamp')) {
      const now = new Date();
      fieldsList.push(now.toISOString());
    }

    if (CONFIG.error_log_fields.includes('level')) {
      fieldsList.push(level);
    }

    if (CONFIG.error_log_fields.includes('context')) {
      const callStackCallerFuncPos = 2;
      fieldsList.push(new Error().stack.split('\n')[callStackCallerFuncPos]);
    }

    if (CONFIG.error_log_fields.includes('var_name')) {
      fieldsList.push('var_name' in fields ? fields.var_name : CONFIG.error_log_empty_field);
    }

    if (CONFIG.error_log_fields.includes('var_value')) {
      fieldsList.push('var_value' in fields ? `${fields.var_value} /${typeof fields.var_value}/` : CONFIG.error_log_empty_field);
    }

    if (CONFIG.error_log_fields.includes('msg')) {
      fieldsList.push('msg' in fields ? fields.msg : CONFIG.error_log_empty_field);
    }

    console.error(fieldsList.join(CONFIG.error_log_field_sep));
  }
};

const access = (fields) => {
  if (CONFIG.access_log_enabled && accessLogAvailable) {
    const fieldsList = [];

    if (CONFIG.access_log_fields.includes('pid')) {
      fieldsList.push(process.pid);
    }

    if (CONFIG.access_log_fields.includes('timestamp')) {
      const now = new Date();
      fieldsList.push(now.toISOString());
    }

    if (CONFIG.access_log_fields.includes('remote_addr')) {
      fieldsList.push('remote_addr' in fields ? fields.remote_addr : CONFIG.access_log_empty_field);
    }

    if (CONFIG.access_log_fields.includes('req_line')) {
      fieldsList.push('req_line' in fields ? fields.req_line : CONFIG.access_log_empty_field);
    }

    if (CONFIG.access_log_fields.includes('user_agent')) {
      fieldsList.push('user_agent' in fields ? fields.user_agent : CONFIG.access_log_empty_field);
    }

    if (CONFIG.access_log_fields.includes('status_code')) {
      fieldsList.push('status_code' in fields ? fields.status_code : CONFIG.access_log_empty_field);
    }

    if (CONFIG.access_log_fields.includes('content_length')) {
      fieldsList.push('content_length' in fields ? fields.content_length : CONFIG.access_log_empty_field);
    }

    fieldsList.push('\n');

    const accessLogEntry = fieldsList.join(CONFIG.access_log_field_sep);

    const canWriteMore = accessLog.write(accessLogEntry, 'utf-8', () => {
      error(errorLogLevels.DEBUG, { msg: 'line written to access log' });
    });

    if (!canWriteMore) {
      error(errorLogLevels.ERROR, { msg: 'access log write stream buffer filled. logging temporarily turned off. logs might be missing' });
      accessLogAvailable = false;
      accessLog.on('drain', () => {
        error(errorLogLevels.ERROR, { msg: 'access log write stream buffer drained. logging turned on.' });
        accessLogAvailable = true;
        accessLog.removeAllListeners('drain');
      });
    }
  }
};

module.exports = {
  log: {
    error,
    access,
  },
  errorLogLevels,
};
